src/js_sync.hpp

```diff
@@ -62,11 +64,18 @@ static std::string partition_value_bson_to_string(typename T::Context ctx, typen
         bson::Bson partition_bson;
         if (Value<T>::is_string(ctx, partition_value_value)) {
             std::string pv = Value<T>::validated_to_string(ctx, partition_value_value);
+            if (pv.length() == 0) {
+                throw std::runtime_error("partitionValue of type 'string' may not be an empty string.");
+            }
             partition_bson = bson::Bson(pv);
         }
         else if (Value<T>::is_number(ctx, partition_value_value)) {
             auto pv = Value<T>::validated_to_number(ctx, partition_value_value);
-            partition_bson = bson::Bson(static_cast<int64_t>(pv));
+            if (pv < 0.0 || pv > JS_MAX_SAFE_INTEGER || fmod(pv, 1.0) > 0.0) {
+                throw std::runtime_error("partitionValue of type 'number' must be a non-negative integer <= Number.MAX_SAFE_INTEGER.");
+            }
+            auto pvi = static_cast<int64_t>(pv);
```

Comment:
I would use `modf` and cast the integral part instead of casting pv. 
http://www.cplusplus.com/reference/cmath/modf/
```
double integerPart;
fractionalPart = modf(value, &integerPart);
```
and compare using the `fractionalPart > 0.0` and cast the `integerPart`

Reply:
@fronck also suggested `modf`. I've updated to this way instead, but wouldn't the result be the same?